src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+            {
+                var document = solution.GetDocument(id);
+                var targetNamespace = _state.TargetNamespace;
+
+                var declarationRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var container = declarationRoot.DescendantNodes().FirstOrDefault(node => node is TNamespaceDeclarationSyntax) ?? declarationRoot;
+
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var declaredSymbols = GetDeclaredSymbolsInContainer(semanticModel, container, cancellationToken);
+
+                var documentSet = ImmutableHashSet.Create(document);
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+                var refLocationsInCurrentDocument = new List<ReferenceLocation>();
+                var refLocationsInOtherDocuments = new List<ReferenceLocation>();
+
+                foreach (var declaredSymbol in declaredSymbols)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var refSymbols = await SymbolFinder.FindReferencesAsync(
+                      declaredSymbol,
+                      solution,
+                      cancellationToken).ConfigureAwait(false);
+
+                    var refLocationsForSymbol = refSymbols.Where(refSymbol => refSymbol.Definition == declaredSymbol)
+                        .SelectMany(refSymbol => refSymbol.Locations);
+
+                    foreach (var refLocation in refLocationsForSymbol)
+                    {
+                        if (refLocation.Document.Id == document.Id)
+                        {
+                            refLocationsInCurrentDocument.Add(refLocation);
+                        }
+                        else
+                        {
+                            Debug.Assert(!PathUtilities.PathsEqual(refLocation.Document.FilePath, document.FilePath));
+                            refLocationsInOtherDocuments.Add(refLocation);
+                        }
+                    }
+                }
+
+                document = await FixDeclarationDocumentAsync(document, refLocationsInCurrentDocument, cancellationToken)
+                    .ConfigureAwait(false);
+                solution = document.Project.Solution;
+
+                var refLocationGroups = refLocationsInOtherDocuments.GroupBy(loc => loc.Document.Id);
+                foreach (var refInOneDocument in refLocationGroups)
+                {
+                    var refDocument = solution.GetDocument(refInOneDocument.Key);
+                    document = await FixReferencingDocumentAsync(refDocument, refInOneDocument, cancellationToken)
+                        .ConfigureAwait(false);
+                    solution = document.Project.Solution;
```

Comment:
is it actualy necessary that you have to do this serially?  Does each file change depend on the changes to the previous files?  If not, then this is expensive if you use semantics at all.  Even if you don't use semantics, if there's no ordering dependency, we should do this work in parallel.

Reply:
Done